build/pkgs/cython/distros/conda.txt

```diff
@@ -1 +1 @@
-cython
+cython==3.0.2
```

Comment:
Wouldn't it be better to exclude the broken 3.0.3 instead of fixing the version (here and in install-requires)? From what I read, the bug should be fixed in 3.0.4.

Reply:
No, we have to expect further breakage in the 3.0.x series

Reply:
Any basis for this expectation? Looking at https://github.com/sagemath/sage/pull/36110, there was no problem upgrading  from 3.0.0 to 0.1 and 0.2.

I think we can always exclude newer versions once they are known to break something...

Reply:
> we can always exclude newer versions once they are known to break something...

I think I have explained this to you several times in the past already. This type of breakage affects releases that we have already made retroactively. As our project does not make bug fix releases, we *cannot* fix such breakage after it happens.

So when our experience indicates that we have to expect breakage from updates, it is prudent to put in suitable version constraints.

The breakage with 3.0.3 affected a major project (scipy), so this is an indication that the Cython project is currently struggling with release management. This is understandable because after the rollout of Cython 3 in distributions, there is likely very high demand for developer attention as lots of projects are forced to add support for Cython 3.

Reply:
> So when our experience indicates that we have to expect breakage from updates, it is prudent to put in suitable version constraints.

Yes, that's why you put in the major version constraint for cython. Bugs in other projects can always happen, but that's not a valid reason to fix the version. (A better solution in the form of a lock file is waiting for review: https://github.com/sagemath/sage/pull/35986)

For example, if we release sage with this version constraint, then it is no longer possible to use it in a conda environment with a newer version of cython. But someone might need 3.0.4 because earlier versions break of them.

So please loosen the constraint again. Thanks

Reply:
> that's not a valid reason to fix the version.

Sorry, Tobias, you don't get to declare that.

Reply:
By that logic you would need to introduce a lot more fixed version constraints for a lot more packages. A lot of our dependencies have weaker ci than cython and thus may introduce new bugs.

For me these version constraints say "we expect that sage is compatible with these versions". Since that range usual contains future releases, it implicitly has the addition "of course, only given that the dependency doesn't introduce breaking changes/bugs". 

From what I can see the cython issue that affects us is likely fixed in 3.0.4 (it already has patches discussed etc). So until you have evidence that there might be more bugs that affect us, I have the expectation that 3.0.4 works.

Reply:
It's not helpful to repeat the same points every time that a version constraint has to be set. I'm fully aware of any and all of these points, and they are already factored in.

Reply:
> I'm fully aware of any and all of these points, and they are already factored in.

Perfect! In that case you might want to point out where our evaluation differs, since apparently we come to a different end result. For example, you might want to start with discussing why we didn't needed a fixed version before but now we do. For all I can see the cython 3 gets more stable and stable, so there are less and less bugs expected. By this logic, we would need a fixed version less and less.